File: python/Data Structures and Algorithms in Python Book/arrays/caesar_cipher.py
class CaesarCipher:
    """Class for encryption & decryption using Caesar cipher"""

    def __init__(self, shift):
        """Construct Caeser cipher using integer shift for rotation"""
        encoder = [None] * 26
        decoder = [None] * 26
        for k in range(26):
            encoder[k] = chr((k+shift) % 26 + ord('A')) 
            decoder[k] = chr((k -shift) % 26 + ord('A'))
        self._forward = ''.join(encoder)                        # will store as string
        self._backward = ''.join(decoder)                       # use it for decoding
    
    def encrypt(self, message):
        """Return encryted message string"""
        return self._transform(message, self._forward)
    
    def decrypt(self, secret_message):
        """Return decryted message string"""
        return self._transform(secret_message, self._backward) 
    
    def _transform(self, original, code):
        msg_char_list = list(original)
        for k in range(len(msg_char_list)):
            if msg_char_list[k].isupper():
                # find index of that char of msg in the enoded/decoded char list
                j = ord(msg_char_list[k]) - ord('A')
                msg_char_list[k] = code[j]
        return ''.join(msg_char_list)
    
if __name__ == "__main__":
    caesar_cipher = CaesarCipher(shift=3)
    message = "THE EAGLE IS IN PLAY; MEET AT JOE S."
    encrypted_message = caesar_cipher.encrypt(message)
    print("Secret:", encrypted_message)
    decrypted_message = caesar_cipher.decrypt(encrypted_message)
    print("Message:", decrypted_message)

File: java/fds/CaesarCipher.java
package fds;

/** Class for doing encryption and decrytion using the Caesar Cipher. */
public class CaesarCipher {
    protected char[] encoder = new char[26]; // Encrytion array
    protected char[] decoder = new char[26]; // Decryption array

    /**
     * Constructor that initailizes the encrytion and decrytion arrays.
     * 
     * @param rotation number of times each character is rotated/shifted from its
     *                 position
     */
    public CaesarCipher(int rotation) {
        for (int k = 0; k < 26; k++) {
            encoder[k] = (char) ('A' + (k + rotation) % 26); // we add 'A' to get the new char at index k
            decoder[k] = (char) ('A' + (k - rotation + 26) % 26);
        }
    }

    /** Returns String representation of the encrypted message. */
    public String encrypt(String message) {
        return transform(message, encoder);
    }

    /** Returns decrypted message given encvrypted secret. */
    public String decrypt(String secret) {
        return transform(secret, decoder);
    }

    /** Returns transformation of original String using the given code. */
    public String transform(String original, char[] code) {
        char[] msg = original.toCharArray();
        for (int k = 0; k < msg.length; k++) {
            if (Character.isUpperCase(msg[k])) { // we have a letter to change
                int j = msg[k] - 'A'; // will be value from 0 to 25
                msg[k] = code[j]; // replace the character
            }
        }
        return new String(msg);
    }

    /** Simple main method for testing the Caesar Cipher. */
    public static void main(String[] args) {
        CaesarCipher cipher = new CaesarCipher(3);
        System.out.println("Encryption code = " + new String(cipher.encoder));
        System.out.println("Decryption code = " + new String(cipher.decoder));
        String message = "THE EAGLE IS IN PLAY; MEET AT JOE'S";
        String coded = cipher.encrypt(message);
        System.out.println("Secret: " + coded);
        String decoded = cipher.decrypt(coded);
        System.out.println("Message: " + decoded);

        cipher = new CaesarCipher(1);
        System.out.println("Encryption code = " + new String(cipher.encoder));
        System.out.println("Decryption code = " + new String(cipher.decoder));
        message = "THE EAGLE IS IN PLAY; MEET AT JOE'S";
        coded = cipher.encrypt(message);
        System.out.println("Secret: " + coded);
        decoded = cipher.decrypt(coded);
        System.out.println("Message: " + decoded);

    }
}